.github/copilot-instructions.md

```diff
@@ -0,0 +1,187 @@
+
+## Repository Structure
+
+- `python/egglog/` - Main Python package source code
+- `python/tests/` - Python test suite (pytest-based)
+- `src/` - Rust source code for Python bindings (PyO3)
+- `docs/` - Documentation source files (Sphinx)
+- `test-data/` - Test data files
+- `pyproject.toml` - Python project configuration and dependencies
+- `Cargo.toml` - Rust project configuration
+- `uv.lock` - Locked dependencies (managed by uv)
+
+## Build and Development Commands
+
+### Prerequisites
+- **uv** - Package manager (https://github.com/astral-sh/uv)
+- **Rust toolchain** - Version 1.79.0 (specified in rust-toolchain.toml)
+- **Python** - 3.10+ (see .python-version)
+
+### Common Commands
+
+```bash
+# Install dependencies
+uv sync --extra dev --locked
+
+# Run tests
+uv run pytest --benchmark-disable -vvv --durations=10
+
+# Type checking with mypy
+make mypy
+
+# Stub testing
+make stubtest
+
+# Build documentation
+make docs
+
+# Format code (auto-run by pre-commit)
+uv run ruff format .
+
+# Lint code (auto-run by pre-commit)
+uv run ruff check --fix .
+```
+
+## Python Code Standards
+
+### General Guidelines
+- **Line length**: 120 characters maximum
+- **Type hints**: Use type annotations for public APIs and functions
+- **Formatting**: Use Ruff for code formatting and linting
+- **Testing**: Write tests using pytest in `python/tests/`
+- **Docstrings**: Use clear, concise docstrings for public functions and classes
+
+### Ruff Configuration
+The project uses Ruff for linting and formatting with specific rules:
+- Allows uppercase variable names (N806, N802)
+- Allows star imports (F405, F403)
+- Allows `exec` and subprocess usage (S102, S307, S603)
+- Allows `Any` type annotations (ANN401)
+- Test files don't require full type annotations
+
+See `pyproject.toml` for complete Ruff configuration.
+
+### Type Checking
+- **mypy** is used for static type checking
+- Run `make mypy` to type check Python code
+- Run `make stubtest` to validate type stubs against runtime behavior
+- Exclusions: `__snapshots__`, `_build`, `conftest.py`
+
+### Testing
+- Tests are located in `python/tests/`
+- Use pytest with snapshot testing (pytest-inline-snapshot)
+- Benchmarks use pytest-benchmark and CodSpeed
+- Run tests with: `uv run pytest --benchmark-disable -vvv`
+
+## Rust Code Standards
+
+### General Guidelines
+- **Edition**: Rust 2024
+- **FFI**: Uses PyO3 for Python bindings
+- **Main library**: Uses egglog from git (currently saulshanabrook/egg-smol branch)
+
+### File Organization
+- `src/lib.rs` - Main library entry point
+- `src/egraph.rs` - E-graph implementation
+- `src/conversions.rs` - Type conversions between Python and Rust
+- `src/py_object_sort.rs` - Python object handling
+- `src/extract.rs` - Extraction functionality
+- `src/error.rs` - Error handling
+- `src/serialize.rs` - Serialization support
+- `src/termdag.rs` - Term DAG operations
+- `src/utils.rs` - Utility functions
```

Comment:
Can you add overviews of the Python files as well?

And document that the public python interface is all from `egglog` so anything that is public should be exported at the top level.

There is also the lower level `egglog.bindings` module which can also be used.